src/components/Rules/Rules.js

```diff
@@ -1,65 +1,49 @@
 import React from 'react';
 import RulesContainer from './RulesContainer'
 import AppBar from '../Shared/AppBar';
+import * as rulesUtils from '../../utils/rulesUtils';
 
 export default class Rules extends React.Component {
   constructor(props) {
     super(props);
     this.onDragEnd = this.onDragEnd.bind(this);
     this.deleteRule = this.deleteRule.bind(this);
     this.addRule = this.addRule.bind(this);
-    this.state = { rules: [], max_id: 0};
+    this.saveChanges = this.saveChanges.bind(this);
+    this.state = { rules: [], max_index: 0, savedRules: [], canSaveChanges: false};
   }
 
-  componentDidMount() {
-    this.getCurrentRules();
+  async componentDidMount() {
+    await this.getCurrentRules();
   }
 
   deleteRule(id) {
-    const rules = this.state.rules.filter(rule => rule['id'] !== id);
+    const deletedRule = this.state.rules.filter(rule => rule['id'] === id)[0];
+    let new_rules = this.state.rules.filter(rule => rule['id'] !== id);
+    new_rules = rulesUtils.updateIndexesFromDeletion(new_rules, deletedRule['index']);
     this.setState({
-      rules: rules
+      rules: new_rules
     });
-  }
-
-  updateIndexes(new_rules) {
-    for (let rule of new_rules) {
-      rule['index'] = rule['index'] + 1;
-    }
-    return new_rules;
+    this.updateSavingButton(new_rules);
   }
 
   addRule(rule) {
-    rule['id'] = this.state.max_id + 1;
-    this.setState({max_id: this.state.max_id + 1})
+    rule['id'] = this.state.max_index + 1;
+    this.setState({max_index: this.state.max_index + 1})
     rule['index'] = 0;
     rule['created'] = true;
     let new_rules = JSON.parse(JSON.stringify(this.state.rules)); //deep clone to update changes
-    new_rules = this.updateIndexes(new_rules);
+    new_rules = rulesUtils.updateIndexesFromAddition(new_rules);
     new_rules.unshift(rule);
     this.setState({
       rules: new_rules
     });
+    this.updateSavingButton(new_rules);
   }
 
-  getCurrentRules() {
-    // We should hit an endpoint to get the current rules
-    const rules = this.mockRules();
-    this.setState({
-      rules: rules
-    });
-  }
-
-  reorder(list, startIndex, endIndex) {
-    const result = Array.from(list);
-    const [removed] = result.splice(startIndex, 1);
-    result.splice(endIndex, 0, removed);
-
-    result.forEach(function (rule, i) {
-      rule.index = i;
-    });
-
-    return result;
+  async getCurrentRules() {
+    const rules = await rulesUtils.getRules();
+    this.setState({ max_index: rules.length, rules: rules, savedRules: JSON.parse(JSON.stringify(rules)) });
```

Comment:
porq lo stringifyas y despues lo parseas? no seria lo mismo q no hacerle nada?

Reply:
Necesitaba hacerle un deep copy al objeto rules para que al menos cambie el id si todos los valores son iguales y me actualice el estado, y vi que esta era la forma más estandar: https://stackoverflow.com/questions/122102/what-is-the-most-efficient-way-to-deep-clone-an-object-in-javascript